Add Snowflake support: introduce SnowflakeDatabaseType

caoterry opened the pull request:
Adds first-class Snowflake support via `SnowflakeDatabaseType`.

Related to goldmansachs/reladomo#264

## What & Why
- Provide a dedicated `SnowflakeDatabaseType` aligned with Snowflake semantics (no row-level locks, session-scoped temp tables, LIMIT syntax, etc.).

## Key changes
- CRUD works on Snowflake with vendor-specific semantics:
  - SELECT: uses `LIMIT (rowCount + 1)` for capped reads; `FOR UPDATE` ignored (no row-level locks).
  - UPDATE: batch/multi updates use `UPDATE … FROM` join with temp tables.
  - DELETE: WHERE-only deletes (Snowflake has no `DELETE … LIMIT`); no bulk-delete temp-table join in this PR.
  - INSERT: supported via batched prepared statements (configurable batch sizes).
- Temp tables: non-shared only — `CREATE TEMPORARY TABLE` (session-scoped). Shared temp tables are not supported on Snowflake.
- Test tables: `shouldCreateTestIndexes() == false` → skip PK/UNIQUE index creation (Snowflake has no traditional indexes).
- LIKE escaping: meta chars `{ '%', '_', '\\' }`.
- Type mappings aligned to Snowflake (`number`, `timestamp_ntz`, `binary`, etc.).

## Tests
- 78 tests pass locally; can add more vendor/Snowflake-specific tests.

## Backward compatibility
- Other database types unchanged.

## Environment
- OS: Windows 11 + WSL2 (Ubuntu 24.04.2 LTS, kernel 6.6.87.2-microsoft-standard-WSL2)
- JDK: OpenJDK 1.8.0_462 (build 1.8.0_462-8u462-ga~us1-0ubuntu2~24.04.2-b08)
- Snowflake JDBC: 3.26.1
- Runner: VS Code Java Debug + JUnit 4 (`org.junit.runner.JUnitCore`)
- JVM flags: `-Duser.timezone=UTC`

.vscode/launch.json:
{
  "type": "java",
  "name": "Debug Single Snowflake Test",
  "request": "launch",
  "mainClass": "org.junit.runner.JUnitCore",
  "args": ["com.gs.fw.common.mithra.test.TestSnowflakeGeneralTestCases"],
  "cwd": "${workspaceFolder}",
  "env": { "RELADOMO_HOME": "${workspaceFolder}" },
  "vmArgs": "-Duser.timezone=UTC"
}

## Tests
<img width="378" height="215" alt="image" src="https://github.com/user-attachments/assets/cb4ad4db-ff96-461e-9712-c1441b49bf96" />

Nov-26:
<img width="1096" height="403" alt="image" src="https://github.com/user-attachments/assets/ac7a8914-6786-4428-9e1f-152dcfcd30f9" />

mohrezaei commented on reladomo/src/main/java/com/gs/fw/common/mithra/databasetype/SnowflakeDatabaseType.java:
```diff
@@ -0,0 +1,619 @@
+    @Override
+    public String getSelectForAggregatedData(SqlQuery query, List aggregateAttributes, List groupByAttributes)
+    {
+        String sql = super.getSelectForAggregatedData(query, aggregateAttributes, groupByAttributes);
```
I don't think this override is reasonable.
some of it just doesn't make sense, for example, casting `count` does nothing because it happens after count is invoked.
Some of it appears to be somewhat application/schema related. If deterministic math is required, shouldn't the column type be `Numeric` to begin with? This is a choice that an application should make, not Reladomo.
For applications where this is potentially problematic because of legacy schemas, this function could be overriden in an app specific databasetype.

Please remove this override.

caoterry replied:
Hi Moh, 

Thank you so much for taking the time to review my code earlier — really appreciate your insights and suggestions.

Apologies for the delayed response. I was out on vacation for two weeks and things have been quite hectic since I returned, but I'm now catching up and working to get the pull request to mergeable state.

Just to share a bit more context: I originally added the override in `getSelectForAggregatedData()` to make certain tests in the vendor test suite pass — in particular, `TestVariance().testVariancePopFloat()` and `TestStandardDeviation().testStdDevPopFloat()` and a few other similar test cases, which were failing under Snowflake due to numeric precision inconsistencies.

I’m now considering whether I should drop the override entirely and remove those related vendor test cases.

I’ll aim to submit an updated version shortly. Thanks again for your feedback!

Many Thanks
Terry

mohrezaei replied:
For numeric precision, feel free to add error bounds to the assert instead of disabling.

caoterry replied:
I’ve removed the extra casts and added explicit error bounds in the `TestSnowflakeGeneralTestCases`.
I’m still keeping a minimal override in `SnowflakeDatabaseType#getSelectForAggregatedData`; the rationale is described in the doc above

mohrezaei commented:
Other than the one comment, this PR looks really good. thanks for the contribution.

caoterry committed:
Add Snowflake support: SnowflakeDatabaseType and Snowflake tests

covered by: tianyi_cao.dco

caoterry commented:
Quick update on this PR:

1. Addressed the earlier review comments (removed the unnecessary cast and tightened up the aggregation logic / error bounds).
2. Squashed the branch into a single commit on top of `upstream/master`.
3. Added the DCO entry and ensured the Apache 2.0 headers are in place.

On the DCO / copyright side:

I’ve added a personal DCO entry (`dco/tianyi_cao.dco`) and used
`covered by: tianyi_cao.dco` in the commit message.

I’ve left the file headers with the existing Goldman Sachs Apache 2.0 copyright
notice only, without adding an extra
`// Portions copyright <my name>. Licensed under Apache 2.0 license`
line, mainly to avoid confusion with my current employer’s internal IP policies.

If you’d prefer explicit per-file attribution or a different convention here,
I’m happy to update the headers to match whatever you recommend.

mohrezaei commented:
Thanks again for the contribution.

caoterry commented:
> Thanks again for the contribution.

Thanks as well! I’ve been using Reladomo for years in Controllers Engineering and really enjoy working with it — very happy to be able to contribute back.

bgshweta commented:
@mohrezaei Can you please help create tag with this change?

mohrezaei commented:
Just a tag or a full release?

mohrezaei commented:
Please also note that you can simply use the database type code without a new release.

bgshweta commented:
How about full release as this new database support addition? But OK to go ahead with a TAG and creating full release after testing it for a while. Need the version to be mirrored to internal maven.

mohrezaei commented:
tag v18.1.1-rc0

bgshweta commented:
Thank you! Will try testing with this and reach out for release tag creation.

bgshweta commented:
@mohrezaei I am unable to find the artifacts for the tag here - https://central.sonatype.com/search?q=g:com.goldmansachs.reladomo&smo=true
Does this need any additional step or do only full releases get published to maven? If it is latter are you ok to create a full release for this change? If not I will have to find way to download the artifacts directly somehow.

mohrezaei commented:
only full releases are pushed to maven.

mohrezaei commented:
I would recommend building the artifacts locally and testing them in your application before a release.

Prat2404 commented:
Hi! Any plans for a new release soon? It would be great to get this published so we can try out the new Snowflake support.
